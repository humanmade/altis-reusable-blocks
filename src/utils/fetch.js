import _zipObject from 'lodash/zipObject';

import apiFetch from '@wordpress/api-fetch';

/**
 * Fetch JSON.
 *
 * Helper function to return parsed JSON and also the response headers.
 *
 * @param {Object} args - Array of arguments to pass to apiFetch.
 * @param {Object[]} headerKeys - Array of headers to include.
 */
export const fetchJson = ( args, headerKeys = [ 'x-wp-totalpages' ] ) => {
	return apiFetch(
		{
			...args,
			parse: false,
		}
	).then( ( response ) => {
		return Promise.all( [
			response.json ? response.json() : [],
			_zipObject( headerKeys, headerKeys.map( key => response.headers.get( key ) ) ),
		] );
	} );
};
