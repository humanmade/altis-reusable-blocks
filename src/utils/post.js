import { sprintf } from '@wordpress/i18n';

import settings from '../settings';

const { editPostUrl } = settings;

/**
 * Return the Edit URL for the post with the given ID.
 *
 * @param {number} postId - Post ID.
 *
 * @return {string} URL.
 */
export function getEditPostUrl( postId ) {
	return sprintf( editPostUrl, postId );
}
