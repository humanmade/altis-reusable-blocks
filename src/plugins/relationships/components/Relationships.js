import _uniqBy from 'lodash/uniqBy';
import PropTypes from 'prop-types';

import {
	PanelBody,
	PanelRow,
	Placeholder,
	Spinner,
} from '@wordpress/components';
import { PluginSidebar, PluginSidebarMoreMenuItem } from '@wordpress/edit-post';
import { Component, Fragment } from '@wordpress/element';
import { __ } from '@wordpress/i18n';
import { addQueryArgs } from '@wordpress/url';

import settings from '../../../settings';
import { fetchJson } from '../../../utils/fetch';

import Pagination from './Pagination';
import RelationshipItem from './RelationshipItem';

const { relationshipsPerPage } = settings;

const baseClassName = 'erb-relationships';
const sidebarName = 'erb-relationships';

class Relationships extends Component {
	state = {
		relationshipsList: [],
		currentPage: 1,
		isFetching: false,
		totalPages: 0,
		totalItems: 0,
	};

	/**
	 * When component mounts, fetch relationships for the block.
	 */
	componentDidMount() {
		this.fetchRelationships();
	}

	/**
	 * Fetch all the posts that use the current block.
	 *
	 * @param {Number} page - Page number for the request.
	 */
	fetchRelationships = async ( page = 1 ) => {
		const { currentPostId } = this.props;

		this.setState( { isFetching: true } );

		try {
			const data = await fetchJson(
				{
					path: addQueryArgs(
						`/altis-erb/v1/relationships`, {
							block_id: currentPostId,
							page,
						}
					),
				},
				[ 'x-wp-totalpages', 'x-wp-total' ]
			);

			this.updateRelationshipsList( data );
		} catch ( e ) {
			/* eslint-disable no-console */
			console.error( 'Error retrieving relationships for block.' );
			console.error( e );
			/* eslint-enable no-console */
		}

		this.setState( { isFetching: false } );
	};

	/**
	 * Update the Relationships List state object with a new list and normalize them.
	 *
	 * @param {Object[]} data - Array of new relationships fetched and response headers.
	 * @param {Object[]} data.newRelationshipsList - Array of new relationships fetched.
	 * @param {Object[]} data.headers - Array of response headers.
	 */
	updateRelationshipsList = ( [ newRelationshipsList, headers ] ) => {
		const { relationshipsList } = this.state;

		if ( ! newRelationshipsList.every( ( item ) => relationshipsList.includes( item ) ) ) {
			const totalPages = parseInt( headers[ 'x-wp-totalpages' ], 10 );
			const totalItems = parseInt( headers[ 'x-wp-total' ], 10 );

			this.setState( {
				relationshipsList: _uniqBy( [ ...relationshipsList, ...newRelationshipsList ], 'id' ),
				totalPages,
				totalItems,
			} );
		}
	};

	/**
	 * Changes crurrentPage state to previous page.
	 */
	goToPrevPage = () => {
		const { currentPage } = this.state;

		if ( currentPage > 1 ) {
			this.setState( {
				currentPage: currentPage - 1,
			} );
		}
	};

	/**
	 * Changes crurrentPage state to next page.
	 */
	goToNextPage = () => {
		const {
			currentPage,
			relationshipsList,
			totalPages,
		} = this.state;

		if ( currentPage < totalPages ) {
			this.setState( {
				currentPage: currentPage + 1,
			} );

			if ( relationshipsList.length <= currentPage * relationshipsPerPage ) {
				this.fetchRelationships( currentPage + 1 );
			}
		}
	};

	render() {
		const {
			currentPage,
			isFetching,
			relationshipsList,
			totalPages,
			totalItems,
		} = this.state;

		const title = __( 'Relationships', 'enhanced-reusable-blocks' );

		const startIndex = currentPage === 1 ? 0 : ( currentPage - 1 ) * relationshipsPerPage;

		const items = relationshipsList.slice( startIndex, relationshipsPerPage * currentPage );
		const relationshipItems = items.length
			? items.map( ( relationshipItem ) => {
				return ( <RelationshipItem { ...relationshipItem } key={ relationshipItem.id } /> );
			} )
			: <PanelRow>{ __( 'No Relationships to Display', 'enhanced-reusable-blocks' ) }</PanelRow>;

		return (
			<Fragment>
				<PluginSidebarMoreMenuItem target={ sidebarName }>
					{ title }
				</PluginSidebarMoreMenuItem>
				<PluginSidebar
					name={ sidebarName }
					title={ title }
				>
					<PanelBody className={ `${ baseClassName }__relationships_list` }>
						{
							isFetching
								? <Placeholder><Spinner /></Placeholder>
								: relationshipItems
						}
						<Pagination
							currentPage={ currentPage }
							goToPrevPage={ this.goToPrevPage }
							goToNextPage={ this.goToNextPage }
							totalPages={ totalPages }
							totalItems={ totalItems }
						/>
					</PanelBody>
				</PluginSidebar>
			</Fragment>
		);
	}
}

Relationships.propTypes = {
	currentPostId: PropTypes.number.isRequired,
};

export default Relationships;
