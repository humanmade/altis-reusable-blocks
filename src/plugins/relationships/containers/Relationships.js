import { compose } from '@wordpress/compose';
import { withSelect } from '@wordpress/data';

import Relationships from '../components/Relationships';

export const mapSelectToProps = ( select ) => {
	const { getCurrentPostId } = select( 'core/editor' );

	return {
		currentPostId: getCurrentPostId(),
	};
};

export default compose( [
	withSelect( mapSelectToProps ),
] )( Relationships );
