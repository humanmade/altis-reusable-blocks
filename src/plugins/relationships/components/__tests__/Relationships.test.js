import { shallow } from 'enzyme';

import { addQueryArgs } from '@wordpress/url';

import Relationships from '../Relationships';

import { fetchJson } from '../../../../utils/fetch';

jest.mock( '@wordpress/url', () => ( { addQueryArgs: jest.fn() } ) );

jest.mock( '../Pagination', () => mockComponent( 'Pagination' ) );

jest.mock( '../RelationshipItem', () => mockComponent( 'RelationshipItem' ) );

jest.mock( '../../../../settings', () => ( {
	relationshipsPerPage: 10,
} ) );

jest.mock( '../../../../utils/fetch', () => ( {
	fetchJson: jest.fn( () => Promise.resolve( [] ) ),
} ) );

afterEach( () => {
	jest.clearAllMocks();
} );

const createRelationshipsList = ( length, start = 1 ) => Array.from( { length }, ( _, i ) => ( { id: start + i } ) );

describe( '<Relationships />', () => {

	test( 'initializes state as expected', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		expect( result.state() ).toMatchSnapshot();
	} );

	test( 'fetches relationships on mount', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const instance = result.instance();

		instance.fetchRelationships = jest.fn();

		instance.componentDidMount();

		expect( instance.fetchRelationships ).toHaveBeenCalled();
	} );

	test( 'fetches relationships', async () => {
		const data = [ createRelationshipsList( 3 ), [] ];

		fetchJson.mockImplementationOnce( () => Promise.resolve( data ) );

		const currentPostId = 42;

		const result = shallow(
			<Relationships currentPostId={ currentPostId } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const instance = result.instance();

		instance.updateRelationshipsList = jest.fn();

		await instance.fetchRelationships();

		expect( addQueryArgs ).toHaveBeenCalledWith( '/altis-reusable-blocks/v1/relationships', {
			block_id: currentPostId,
			page: 1,
		} );
		expect( instance.updateRelationshipsList ).toHaveBeenCalledWith( data );
		expect( result.state( 'isFetching' ) ).toBe( false );
	} );

	test( 'handles rejection while fetching relationships', async () => {
		fetchJson.mockImplementationOnce( () => Promise.reject( new Error() ) );

		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const instance = result.instance();

		instance.updateRelationshipsList = jest.fn();

		await instance.fetchRelationships();

		expect( instance.updateRelationshipsList ).not.toHaveBeenCalled();
		expect( result.state( 'isFetching' ) ).toBe( false );
	} );

	test( 'handles error while fetching relationships', async () => {
		fetchJson.mockImplementationOnce( () => {
			throw new Error();
		} );

		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const instance = result.instance();

		instance.updateRelationshipsList = jest.fn();

		await instance.fetchRelationships();

		expect( instance.updateRelationshipsList ).not.toHaveBeenCalled();
		expect( result.state( 'isFetching' ) ).toBe( false );
	} );

	test( 'updating relationships list with empty data does not overwrite existing data', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const state = {
			relationshipsList: createRelationshipsList( 3 ),
			totalItems: 3,
			totalPages: 1,
		};

		result.setState( state );

		result.instance().updateRelationshipsList( [ [] ] );

		expect( result.state() ).toMatchObject( state );
	} );

	test( 'updates relationships list', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const state = {
			relationshipsList: createRelationshipsList( 2 ),
			totalItems: 2,
			totalPages: 1,
		};

		result.setState( state );

		result.instance().updateRelationshipsList( [
			createRelationshipsList( 2, 3 ),
			{
				'x-wp-total': 4,
				'x-wp-totalpages': 2,
			},
		] );

		expect( result.state() ).toMatchObject( {
			relationshipsList: createRelationshipsList( 4 ),
			totalItems: 4,
			totalPages: 2,
		} );
	} );

	test( 'updating relationships list does not add duplicates', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		const state = {
			relationshipsList: createRelationshipsList( 3 ),
			totalItems: 2,
			totalPages: 1,
		};

		result.setState( state );

		result.instance().updateRelationshipsList( [
			createRelationshipsList( 3, 2 ),
			{
				'x-wp-total': 4,
				'x-wp-totalpages': 2,
			},
		] );

		expect( result.state() ).toMatchObject( {
			relationshipsList: createRelationshipsList( 4 ),
			totalItems: 4,
			totalPages: 2,
		} );
	} );

	test( 'navigates to previous page', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( { currentPage: 2 } );

		result.instance().goToPrevPage();

		expect( result.state( 'currentPage' ) ).toBe( 1 );
	} );

	test( 'does not navigate back from first page', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( { currentPage: 1 } );

		result.instance().goToPrevPage();

		expect( result.state( 'currentPage' ) ).toBe( 1 );
	} );

	test( 'navigates to next page', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( {
			currentPage: 1,
			relationshipsList: createRelationshipsList( 20 ),
			totalPages: 2,
		} );

		const instance = result.instance();

		instance.fetchRelationships = jest.fn();

		instance.goToNextPage();

		expect( result.state( 'currentPage' ) ).toBe( 2 );
		expect( instance.fetchRelationships ).not.toHaveBeenCalled();
	} );

	test( 'navigates to new page and fetches items', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( {
			currentPage: 1,
			relationshipsList: createRelationshipsList( 10 ),
			totalPages: 2,
		} );

		const instance = result.instance();

		instance.fetchRelationships = jest.fn();

		instance.goToNextPage();

		expect( result.state( 'currentPage' ) ).toBe( 2 );
		expect( instance.fetchRelationships ).toHaveBeenCalled();
	} );

	test( 'does not navigate beyond last page', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( {
			currentPage: 1,
			totalPages: 1,
		} );

		result.instance().goToNextPage();

		expect( result.state( 'currentPage' ) ).toBe( 1 );
	} );

	test( 'renders as expected if fetching', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( { isFetching: true } );

		expect( result ).toMatchSnapshot();
	} );

	test( 'renders as expected with no items', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		expect( result ).toMatchSnapshot();
	} );

	test( 'renders as expected with items', () => {
		const result = shallow(
			<Relationships currentPostId={ 42 } />,
			{
				disableLifecycleMethods: true,
			}
		);

		result.setState( {
			relationshipsList: createRelationshipsList( 3 ),
			totalItems: 3,
			totalPages: 1,
		} );

		expect( result ).toMatchSnapshot();
	} );

} );
