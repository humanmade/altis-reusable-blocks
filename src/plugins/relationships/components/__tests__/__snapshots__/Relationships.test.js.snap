// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Relationships /> initializes state as expected 1`] = `
Object {
  "currentPage": 1,
  "isFetching": false,
  "relationshipsList": Array [],
  "totalItems": 0,
  "totalPages": 0,
}
`;

exports[`<Relationships /> renders as expected if fetching 1`] = `
<Fragment>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))
    target="relationships"
  >
    Relationships
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))
    name="relationships"
    title="Relationships"
  >
    <ForwardRef(PanelBody)
      className="altis-reusable-block-relationships__relationships_list"
    >
      <Placeholder>
        <Spinner />
      </Placeholder>
      <Pagination
        currentPage={1}
        goToNextPage={[Function]}
        goToPrevPage={[Function]}
        totalItems={0}
        totalPages={0}
      />
    </ForwardRef(PanelBody)>
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))>
</Fragment>
`;

exports[`<Relationships /> renders as expected with items 1`] = `
<Fragment>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))
    target="relationships"
  >
    Relationships
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))
    name="relationships"
    title="Relationships"
  >
    <ForwardRef(PanelBody)
      className="altis-reusable-block-relationships__relationships_list"
    >
      <RelationshipItem
        id={1}
        key="1"
      />
      <RelationshipItem
        id={2}
        key="2"
      />
      <RelationshipItem
        id={3}
        key="3"
      />
      <Pagination
        currentPage={1}
        goToNextPage={[Function]}
        goToPrevPage={[Function]}
        totalItems={3}
        totalPages={1}
      />
    </ForwardRef(PanelBody)>
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))>
</Fragment>
`;

exports[`<Relationships /> renders as expected with no items 1`] = `
<Fragment>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))
    target="relationships"
  >
    Relationships
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebarMoreMenuItem)))>
  <WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))
    name="relationships"
    title="Relationships"
  >
    <ForwardRef(PanelBody)
      className="altis-reusable-block-relationships__relationships_list"
    >
      <PanelRow>
        No Relationships to Display
      </PanelRow>
      <Pagination
        currentPage={1}
        goToNextPage={[Function]}
        goToPrevPage={[Function]}
        totalItems={0}
        totalPages={0}
      />
    </ForwardRef(PanelBody)>
  </WithPluginContext(WithSelect(WithDispatch(PluginSidebar)))>
</Fragment>
`;
